### ISSUE TITLE
False positives with 933160 PL1 PHP Injection Attack: High-Risk PHP Function Call Found

### ISSUE THREAD

**Issue by dune73** [coreruleset/coreruleset#3931]
The quantitative testing project at the CRS dev retreat in Nov 2024 (https://github.com/coreruleset/coreruleset/wiki/Discussion-Quantitative-Testing) revealed some false positives on 933160.

Here is a custom text corpus with the payloads in question: [custom-corpus-933160.txt](https://github.com/user-attachments/files/17628970/custom-corpus-933160.txt)

Here is how to run the corpus against the sandbox:

```
$ cat /tmp/custom-corpus-933160.txt | while read LINE; do PAYLOAD=$(echo $LINE | sed -e "s/^[^\t]*\t//"); echo ; echo $LINE; curl -H "x-format-output: txt-matched-rules" http://sandbox.coreruleset.org/ -d "payload=$PAYLOAD"; done

...

2 It used to be the administrative seat of the Bolewa people in the 16th century and of historical significance of Daniski hill today are the ruins of emirs’ palace, graveyards, old storage system (Silos) and an ancient well.
933160 PL1 PHP Injection Attack: High-Risk PHP Function Call Found
949110 PL1 Inbound Anomaly Score Exceeded (Total Score: 5)
980170 PL1 Anomaly Scores: (Inbound Scores: blocking=5, detection=5, per_pl=5-0-0-0, threshold=5) - (Outbound Scores: blocking=0, detection=0, per_pl=0-0-0-0, threshold=4) - (SQLI=0, XSS=0, RFI=0, LFI=0, RCE=0, PHPI=5, HTTP=0, SESS=0, COMBINED_SCORE=5)

3 It will be joined by Faster Payments and the Cheque & Credit Clearing Company, which incorporates the Image Clearing System (ICS) that will replace the paper processing system for cheques, bringing these major retail payments under one consolidated roof.
933160 PL1 PHP Injection Attack: High-Risk PHP Function Call Found
949110 PL1 Inbound Anomaly Score Exceeded (Total Score: 5)
980170 PL1 Anomaly Scores: (Inbound Scores: blocking=5, detection=5, per_pl=5-0-0-0, threshold=5) - (Outbound Scores: blocking=0, detection=0, per_pl=0-0-0-0, threshold=4) - (SQLI=0, XSS=0, RFI=0, LFI=0, RCE=0, PHPI=5, HTTP=0, SESS=0, COMBINED_SCORE=5)

4 Spruance-class destroyers were all built with the Mark 16 Mod 7 ASROC Launching Group and MK 4 ASROC Weapons Handling System (AWHS) reload system.
933160 PL1 PHP Injection Attack: High-Risk PHP Function Call Found
949110 PL1 Inbound Anomaly Score Exceeded (Total Score: 5)
980170 PL1 Anomaly Scores: (Inbound Scores: blocking=5, detection=5, per_pl=5-0-0-0, threshold=5) - (Outbound Scores: blocking=0, detection=0, per_pl=0-0-0-0, threshold=4) - (SQLI=0, XSS=0, RFI=0, LFI=0, RCE=0, PHPI=5, HTTP=0, SESS=0, COMBINED_SCORE=5)


```

**Comment by dune73** [comment]
[custom-corpus-933160.txt](https://github.com/user-attachments/files/17645392/custom-corpus-933160.txt)

**Comment by franbuehler** [comment]
I tested the false positives from the corpus file and these are the matches:

1. Matched Data: `STAT (\\xd8\\...` found
2. Matched Data: `system (Silos)` found
3. Matched Data: `System (ICS)` found
4. Matched Data: `System (AWHS)` found
5. Matched Data: `file (digunakan oleh JavaScript demonstrasi)` found
6. Matched Data: `system (OCS)` found
7. Matched Data: `System (SLS)` found
8. Matched Data: `System (OCS)` found
9. Matched Data: `System (GNSS)` found
10. Matched Data: `System (JATS)` found
11. Matched Data: `file (misalnya, C    , JavaScript atau file XUL)` found
12. Matched Data: `System (FEDS)` found
13. Matched Data: `System (Sistema di selezione dei modelli a previsione dinamica)` found
14. Matched Data: `System (ECRS)` found
15. Matched Data: `file (non 20 come di consueto)` found

**Comment by theseion** [comment]
The rule could be made more robust by only looking for known shell commands in case of `system`. Also, for all three we could make detection more robust by looking for file paths / names. As these are fairly specific, we might want to create separate rules.

I don't think commenting `system` is a solution, as it's such a simple attack vector.

**Comment by abudawud** [comment]
This is match too: `Understanding & Implementing The Quality Management System (SNI ISO 9001 : 2015)`

I decide to disable this rule on AFTER RESPONSE rule

**Comment by franbuehler** [comment]
Can we remove or comment out system from [933160](https://github.com/coreruleset/coreruleset/blob/main/regex-assembly/933160.ra#L53) when we add it to [unix-shell-evasion-prefix.ra](https://github.com/coreruleset/coreruleset/blob/main/regex-assembly/include/unix-shell-evasion-prefix.ra)?

**Comment by theseion** [comment]
Technically, yes. But if you decide that you don't need the Unix rules because your on Windows... Can `system` be used as an alias in Powershell?

> There are two parts in this file. System is already in the second part, where a regex match is performed to avoid false positives. I don't understand what is meant by this. And adding file to this second part doesn't resolve the fp with the file command.
> So the only solution I currently see is commenting out the two commands file and system.

This refers to `@pmFromFile` in rule `933150`. Those command words used to be part of `933150` but were moved to `933160` to reduce false positives, as the regular expression is much more strict that the substring match performed by `@pmFromFile`.

A pretty simple fix for `system` specifically could be to check for quotes, `$`, `(`+`)`, or `<<<` inside of the parentheses. Since the argument to system must be a string, these are basically the possible options:
- `system("some shell code")`
- `system('more shell code')`
- `system($string_var_evasion)`
- `system(function_call())`
- Heredoc evasion

I'm not sure whether this is legal, but in theory it might be possible to use an inline heredoc to evade detection:
```
system(<<<EOD
heredoc evasion
EOD
)
```

I'd put this into another rule. There are probably other commands that would fall into the same category (`file`, for example).

**Comment by HenningerIBC** [comment]
933160 is still suffering from false positives, especially with the word/function name system.
Isn't the issue that it is just too broad? The regex itself never matches a full function call, as it never checks for a semicolon after the parentheses.

If we check for the semicolon after the parentheses, most false positives should be eliminated. Any reason why this check was never included in this rule?

**Comment by fzipi** [comment]
Confirmed this still reproduces on current `main` (tested against `modsec2-apache`): `system (Silos)` in the corpus payload triggers `933160` — `Matched Data: system (Silos) found within ARGS:payload`. None of the 933160-related fixes merged since this was reported (ReDoS cleanup in #4666, dot-star removal in #4167, unix command list update in #4446) touched the actual cause.

**Root cause**: the rule's mandatory suffix is just `\([^)]*\)` (see `regex-assembly/933160.ra`) — any of the ~15 command words (`system`, `file`, `stat`, `exp`, `ord`, `substr`, `unlink`, `intval`, `prev`, `unserialize`, ...) followed by literal parens containing *anything but a closing paren* matches. Several of those are ordinary English words/abbreviations, so `"System (ICS)"` or `"file (misalnya...)"` in prose match exactly as well as a real call.

**On @theseion's proposal** (2025-02-25) to require a quote, `$`, nested `()`, or `<<<` heredoc marker inside the parens for `system`: I checked it against the existing test suite and it conflicts with an existing required true positive — `tests/regression/tests/REQUEST-933-APPLICATION-ATTACK-PHP/933160.yaml` test_id 25, `system(whoami)`, a bareword argument with none of those markers. So that specific heuristic can't fully separate "real call" from "English phrase" without also giving up detection of bareword-argument calls, which the suite currently requires to match.

(Separately, the most recent suggestion here — requiring a trailing semicolon — would introduce false negatives, since PHP allows omitting the final statement's semicolon before `?>`, e.g. `<?php system($_GET[1])?>`.)

So this looks like a genuine PL1/PL2 tradeoff rather than a pure bug: tightening `system`/`file` (and possibly the other English-word-like commands in the same list) cuts the false positives reported here, but only by dropping the bareword-argument case from PL1 detection — which would need to either move to a higher paranoia level or be accepted as a coverage loss. Flagging for a maintainer decision on scope (which commands, which PL) before anyone implements a fix.

---

**ai disclosure**: Claude (Sonnet 5) — assisted with: reproducing the reported false positive, tracing the root cause in `regex-assembly/933160.ra`, and checking a prior proposed fix against the existing test suite. review performed: reproduced the payload against a live `modsec2-apache` container; read the `.ra` file and `933160.yaml` test file directly to confirm the test_id 25 conflict rather than inferring it.

**Comment by theseion** [comment]
`system(whoami)` is not actually valid PHP syntax. `whoami` here would have to be some symbol. So I think that test case might actually be wrong. Same thing in Python, btw.

**Comment by HackingRepo** [comment]
yes i agree with @theseion, the syntax is invalid in both languages

**Comment by fzipi** [comment]
Good catch, and worth checking rather than taking on faith either way — I looked it up on php.net's own PHP 8.0 migration guide rather than going from memory:

> Attempting to access unqualified constants which are undefined. Previously, unqualified constant accesses resulted in a warning and were interpreted as strings.

So on PHP 8.0+ (released Nov 2020), `whoami` as a bareword throws a fatal `Error` — confirmed locally on PHP 8.5, it's not a warning anymore, execution stops. You're right that on *current* PHP, `system(whoami)` in test_id 25 isn't a working call.

But before PHP 8.0 (i.e. PHP ≤7.4, still running in plenty of legacy/unmaintained installs), an undefined constant silently coerced to its own name as a string with just a warning — so `system(whoami)` on PHP 7.x genuinely did execute `system("whoami")`. That's the well-known "bareword constant" injection trick from that era.

So this doesn't fully resolve the tradeoff, it reframes it: the question isn't "is test 25 testing something that was ever valid," it's "does CRS still want PL1 to catch the bareword-coercion trick against PHP ≤7.4 targets." If the answer is no (only defend currently-supported PHP), test 25 can be dropped/updated and @theseion's quote/`$`/`()`/heredoc check can go in at PL1 without a real coverage loss. If the answer is yes (WAFs often exist specifically to protect the long tail of unpatched/EOL runtimes), it stays a PL1/PL2 split as I described before. Either way, that's a policy call, not something the regex itself can resolve.

---

**ai disclosure**: Claude (Sonnet 5) — assisted with: researching PHP's undefined-constant behavior across versions. review performed: tested the claim locally against PHP 8.5 CLI and cross-checked it against the official php.net PHP 8.0 migration guide before posting, rather than relying on memory.

**Comment by theseion** [comment]
OMFG. 😭

We should comment those rules / regular expressions then, so we will know in the future that these target older PHP versions. We should definitely keep the rules though.

Maybe we can restrict those bare word injections to known shell commands. We already have lists of those commands. That way we could build a rule that better handles the bare word case and a separate one for generic `system` calls.
